File: greedy/11/my.py
class Solution:
    def maxArea(self, height: list[int]) -> int:
        if not height or len(height) < 2:
            return 0
            
        left = 0
        right = len(height) - 1
        max_area = 0
        
        while left < right:
            # Calculate width between lines
            width = right - left
            
            # Height is limited by the shorter line
            curr_height = min(height[left], height[right])
            
            # Calculate current area
            area = width * curr_height
            
            # Update max_area if current area is larger
            max_area = max(max_area, area)
            
            # Move the pointer of the shorter line inward
            # Since keeping the shorter line won't give us a larger area
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
                
        return max_area
    
# In this solution:
# Use two pointers, left and right, starting from the ends of the array
# Calculate the area between the lines at these pointers:

# Width = right - left
# Height = minimum of the two line heights
# Area = width * height

File: greedy/11/my.c
#include <stdio.h>

// Returns the minimum of two integers
int min(int a, int b) {
    return (a < b) ? a : b;
}

// Returns the maximum of two integers
int max(int a, int b) {
    return (a > b) ? a : b;
}

int maxArea(int* height, int heightSize) {
    if (height == NULL || heightSize < 2) {
        return 0;
    }
    
    int left = 0;
    int right = heightSize - 1;
    int maxArea = 0;
    
    while (left < right) {
        // Calculate width between lines
        int width = right - left;
        
        // Height is limited by the shorter line
        int currentHeight = min(height[left], height[right]);
        
        // Calculate current area
        int area = width * currentHeight;
        
        // Update maxArea if current area is larger
        maxArea = max(maxArea, area);
        
        // Move the pointer of the shorter line inward
        // Since keeping the shorter line won't give us a larger area
        if (height[left] < height[right]) {
            left++;
        } else {
            right--;
        }
    }
    
    return maxArea;
}